### ISSUE TITLE
Add undo/redo shortcuts to navigate reflog

### ISSUE THREAD

**Issue by texastoland** [ForkIssues/Tracker#351]
From GitUp and GitKraken. Tower added a dedicated panel.

**Comment by DanPristupov** [comment]
Undo/redo and reflog are kind of different features.

Undo/redo is a quite complicated feature, especially if we talk about rebase or merge.
Reflog can be introduced without any problem.

**Comment by texastoland** [comment]
Maybe I misunderstand reflog? A couple days I rebased without a backup branch and changed my mind. I just reset to one of the most recent commits in terminal.

**Comment by nebhale** [comment]
This matches a workflow I (unfortunately) use occasionally.  It's not actually undo/redo, it's more a way to look back at a huge history of commits (anything that hasn't been GC'd) and arbitrarily point at one of them.  Effectively there's a "trash can" of data that is cleaned up periodically, but not instantly.

It's often used as an undo in that a commit that no longer has a pointer to it, can be resurrected with a new branch or tag.

**Comment by DanPristupov** [comment]
I like the @nebhale's approach. I think reflog should look something like following. In this way it is both useful and won't confuse experienced git users. What do you think guys?
https://user-images.githubusercontent.com/618115/43127975-17b05d44-8f31-11e8-9126-3504eecfeeba.gif (6MB gif)

**Comment by texastoland** [comment]
👏🏽 wow I didn't even consider that!

**Comment by DanPristupov** [comment]
Great! Let me know if you want to have it now and I'll make a pre-release build. Otherwise it will be part of 1.0.70.

**Comment by nebhale** [comment]
I can wait for the first pre-release build to check it out for bugs, but I'm pretty confident that @DanPristupov will get it right on the first try.

**Comment by nebhale** [comment]
Beautiful!  The new lane-shift behavior looks good as well.

![screen shot 2018-07-27 at 10 47 33](https://user-images.githubusercontent.com/60754/43338200-8c6e4bae-918a-11e8-949a-c688115a0109.png)

**Comment by svoop** [comment]
@DanPristupov This is very useful to find commits which still have a parent. However, if I'm not mistaking, the reflog may also contain commits which have lost their parent (e.g. because the parent has been overwritten by a rebase). Does Fork visualize those commits as well somehow?

**Comment by DanPristupov** [comment]
@svoop 

> parent has been overwritten by a rebase

parent can't be **overwritten** because git is immutable. A commit can only be removed after a GC run.

> Does Fork visualize those commits as well somehow?

Yes, enabling reflog after rebase will show the previous state(s).

Update: may be you need `git fsck` terminal command, it shows all lost objects (even the objects which are not connected to the commit tree). Fork doesn't have an UI for it.

**Comment by svoop** [comment]
@DanPristupov 

> parent can't be overwritten because git is immutable. A commit can only be removed after a GC run

You're right of course, what I meant with "overwritten" is not the commit itself, but the commit hash. In my understanding, rebasing a branch will cause all local commits sitting on top to get new commit hashes which means any commit referencing one of those would lose its parent and thus become dangling.

> Update: may be you need git fsck terminal command, it shows all lost objects (even the objects which are not connected to the commit tree). Fork doesn't have an UI for it.

And here they are :smile: Get it, I'm mixing up the reflog and dangling commits. Good to know, thanks for the heads-up!
